agroal-narayana/src/main/java/io/agroal/narayana/LocalXAResource.java

```diff
@@ -36,64 +40,64 @@ public Object getConnection() throws Throwable {
     public void start(Xid xid, int flags) throws XAException {
         if ( currentXid == null ) {
             if ( flags != TMNOFLAGS ) {
-                throw new XAException( "Starting resource with wrong flags" );
+                throw xaException( XAER_RMFAIL, "Starting resource with wrong flags" );
             }
             try {
                 transactionAware.transactionStart();
             } catch ( Exception t ) {
                 transactionAware.setFlushOnly();
-                throw new XAException( "Error trying to start local transaction: " + t.getMessage() );
+                throw xaException( XAER_RMFAIL, "Error trying to start local transaction: " + t.getMessage(), t );
             }
             currentXid = xid;
         } else {
             if ( flags != TMJOIN && flags != TMRESUME ) {
-                throw new XAException( XAException.XAER_DUPID );
+                throw xaException( XAException.XAER_DUPID, "Invalid flag for join|resume" );
             }
         }
     }
 
     @Override
     public void commit(Xid xid, boolean onePhase) throws XAException {
         if ( xid == null || !xid.equals( currentXid ) ) {
-            throw new XAException( "Invalid xid to transactionCommit" );
+            throw xaException( XAER_NOTA, "Invalid xid to transactionCommit" );
         }
 
         currentXid = null;
         try {
             transactionAware.transactionCommit();
         } catch ( Exception t ) {
             transactionAware.setFlushOnly();
-            throw new XAException( "Error trying to transactionCommit local transaction: " + t.getMessage() );
+            throw xaException( onePhase ? XA_RBROLLBACK : XAER_RMFAIL, "Error trying to transactionCommit local transaction: " + t.getMessage(), t );
```

Comment:
XAER_RMFAIL leads to HeuristicRollbackException, while XA_RBROLLBACK leads to RollbackException. I believe the former makes little sense when committing a single resource, therefore opted for the latter when in single phase.